### PR TITLE
Multiple file output

### PR DESCRIPTION
**Pull request by darrell-roberts**
Solution to #71

Adds option to generate multiple modules per source crate.

Supports imports for multiple modules
* Typescript
* Kotlin
* Swift

Updates package declarations for
* Kotlin

Ex:

```
typeshare ~/projects/core --lang typescript --output-folder ~/projects/core/new-types/src --config-file ~/projects/core/typeshare.toml
```

### CODE REVIEW

**Review by CerulanLumina — CHANGES_REQUESTED** [review]
Note: This is a partial review, leaving the following comments. Will return.

File: cli/src/config.rs [hunk 1]
```diff
@@ -63,18 +64,16 @@ pub(crate) struct Config {
     pub go: GoParams,
 }
 
-pub(crate) fn store_config(config: &Config, file_path: Option<&str>) -> Result<(), io::Error> {
+pub(crate) fn store_config(config: &Config, file_path: Option<&str>) -> anyhow::Result<()> {
     let file_path = file_path.unwrap_or(DEFAULT_CONFIG_FILE_NAME);
+    let config_output = toml::to_string_pretty(config).context("Failed to serialize to toml")?;
```
[the remark this answers is not in the document]
Comment (darrell-roberts) [reply cli/src/config.rs:69]: I noticed I had not bumped `toml` along with other deps. I moved it up to `0.8` and it works once again.

File: cli/src/main.rs [hunk 2]
```diff
@@ -170,156 +46,137 @@ fn main() {
...
+    let mut types = TypesBuilder::new();
+    types
+        .add("rust", "*.rs")
+        .context("Failed to add rust type extensions")?;
+    types.select("rust");
+
+    // This is guaranteed to always have at least one value by the clap configuration
+    let first_root = directories
+        .next()
+        .ok_or_else(|| anyhow::Error::msg("directories is empty"))?;
```
[the remark this answers is not in the document]
Comment (darrell-roberts) [reply cli/src/main.rs]: Updated

File: core/src/language/go.rs [hunk 3]
```diff
@@ -26,38 +28,43 @@ pub struct Go {
...
-            types_mapping_to_struct.insert(s.id.original.as_str());
-        }
+        let mut types_mapping_to_struct =
+            HashSet::from_iter(data.structs.iter().map(|s| s.id.original.clone()));
+
         for alias in &data.aliases {
-            if types_mapping_to_struct.contains(&alias.r#type.id()) {
-                types_mapping_to_struct.insert(alias.id.original.as_str());
+            if types_mapping_to_struct.contains(alias.r#type.id()) {
+                types_mapping_to_struct.insert(alias.id.original.clone());
```
Comment (CerulanLumina) [inline core/src/language/go.rs:44]: also here, can we avoid clone / use byref

File: core/src/language/go.rs [hunk 4]
```diff
@@ -26,38 +28,43 @@ pub struct Go {
...
-        for e in &data.enums {
-            items.push(RustItem::Enum(e.clone()))
-        }
+        let mut items = Vec::from_iter(
+            aliases
+                .into_iter()
+                .map(RustItem::Alias)
+                .chain(structs.into_iter().map(RustItem::Struct))
+                .chain(enums.into_iter().map(RustItem::Enum)),
+        );
```
Comment (CerulanLumina) [inline core/src/language/go.rs:63]: minor: Use `.collect` instead of `from_iter`.

File: core/src/language/mod.rs [hunk 5]
```diff
@@ -184,7 +268,16 @@ pub trait Language {
...
+    }
+
+    /// For generating import statements.
+    fn write_imports(
+        &mut self,
+        _writer: &mut dyn Write,
+        _imports: ScopedCrateTypes<'_>,
+    ) -> std::io::Result<()> {
         Ok(())
     }
```
Comment (CerulanLumina) [inline core/src/language/mod.rs:282]: This shouldn't have a default implementation

File: core/src/language/mod.rs [hunk 6]
```diff
@@ -302,4 +395,97 @@ pub trait Language {
...
+                used_imports
+                    .entry(&referenced_import.base_crate)
+                    .and_modify(|names: &mut BTreeSet<&str>| {
+                        names.extend(type_names.iter().map(|s| s.as_str()))
+                    });
+                continue;
+            }
+
+            // Add referenced import for each matching type.
+            if let Some(ty_name) = type_names.get(&referenced_import.type_name) {
```
Comment (CerulanLumina) [inline core/src/language/mod.rs:463]: nit: Normally I like continue but here I think `else if`  better communicates the intent

File: core/src/language/swift.rs [hunk 7]
```diff
@@ -755,6 +768,35 @@ impl Swift {
...
     }
+
+    /// When using mulitple file generation we write this into a separate module vs at the
+    /// end of the generated file.
+    fn write_codable_file(&self, output_folder: &str) -> std::io::Result<()> {
+        let mut w = OpenOptions::new()
+            .write(true)
+            .create(true)
+            .truncate(true)
+            .open(PathBuf::from(output_folder).join("Codable.swift"))?;
```
Comment (CerulanLumina) [inline core/src/language/swift.rs:779]: Can reduce to `File::create`

nit: L779: Can `Path::from` to save 1 aloc

File: core/src/language/mod.rs [hunk 8]
```diff
@@ -302,4 +395,97 @@ pub trait Language {
 
         Ok(())
     }
+
+    /// Types that are remapped will be excluded from import references.
+    fn ignored_reference_types(&self) -> Vec<String> {
+        Vec::new()
+    }
```
Comment (CerulanLumina) [inline core/src/language/mod.rs:402]: nit: can reduce allocs by using `&str`

**Review by CerulanLumina — CHANGES_REQUESTED** [review]
Part 2/2 of review: overall great, some minor notes

File: core/src/language/go.rs [hunk 9]
```diff
@@ -26,38 +28,43 @@ pub struct Go {
...
+        data: ParsedData,
+    ) -> std::io::Result<()> {
         // Generate a list of all types that either are a struct or are aliased to a struct.
         // This is used to determine whether a type should be defined as a pointer or not.
-        let mut types_mapping_to_struct = HashSet::new();
-        for s in &data.structs {
-            types_mapping_to_struct.insert(s.id.original.as_str());
-        }
+        let mut types_mapping_to_struct =
+            HashSet::from_iter(data.structs.iter().map(|s| s.id.original.clone()));
```
[the remark this answers is not in the document]
Comment (darrell-roberts) [reply core/src/language/go.rs:40]: I don't think so, but this is cheaper than the clone that happened before. The parsed data will be moved in each call for `write_enum`, `write_struct` and `write_type_alias` so we cant have references in `types_mapping_to_struct`.  The original code cloned all the types: https://github.com/1Password/typeshare/blob/main/core/src/language/go.rs#L46.

File: core/src/language/go.rs (hunk 3 above)
Comment (darrell-roberts) [reply core/src/language/go.rs:44]: I don't think so, but this is cheaper than the clone that happened before. The parsed data will be moved in each call for `write_enum`, `write_struct` and `write_type_alias` so we cant have references in `types_mapping_to_struct`.  The original code cloned all the types: https://github.com/1Password/typeshare/blob/main/core/src/language/go.rs#L46.

File: core/src/language/go.rs (hunk 4 above)
Comment (darrell-roberts) [reply core/src/language/go.rs]: Updated

File: core/src/language/mod.rs (hunk 5 above)
Comment (darrell-roberts) [reply core/src/language/mod.rs]: Updated

File: core/src/language/mod.rs (hunk 6 above)
Comment (darrell-roberts) [reply core/src/language/mod.rs]: Updated

File: core/src/language/swift.rs (hunk 7 above)
Comment (darrell-roberts) [reply core/src/language/swift.rs]: Updated

File: core/src/language/mod.rs (hunk 8 above)
Comment (darrell-roberts) [reply core/src/language/mod.rs]: Updated

File: cli/src/args.rs [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+        .arg(
+            Arg::new(ARG_OUTPUT_FOLDER)
+                .short('d')
+                .long("output-folder")
+                .help("Folder to write output to. mtime will be preserved if the file contents don't change")
+                .required_unless_present_any([ARG_GENERATE_CONFIG, ARG_OUTPUT_FILE])
+                .takes_value(true)
+                .long(ARG_OUTPUT_FOLDER)
+        )
+        .group(ArgGroup::new("output").args(&["output-file", "output-folder"]).required(true))
```
[the remark this answers is not in the document]
Comment (darrell-roberts) [reply cli/src/args.rs]: Removed required from the ArgGroup and used `.conflicts_with` to make output-folder and output-file mutually exclusive.

File: core/src/language/go.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (darrell-roberts) [reply core/src/language/go.rs:40]: Actually this may be possible. Working on it.

File: core/src/language/go.rs (hunk 3 above)
Comment (darrell-roberts) [reply core/src/language/go.rs:44]: Actually this may be possible. Working on it.

Comment (darrell-roberts) [reply core/src/language/go.rs]: Updated

File: core/src/language/go.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (darrell-roberts) [reply core/src/language/go.rs]: Updated

File: README.md [hunk 11]
```diff
@@ -2,10 +2,10 @@
...
-| typeshare-cli        | [![crates.io version](https://img.shields.io/crates/v/typeshare-cli.svg)](https://crates.io/crates/typeshare-cli) ![MSRV](https://img.shields.io/badge/rustc-1.57+-blue.svg) [![crates.io downloads](https://img.shields.io/crates/d/typeshare-cli.svg)](https://crates.io/crates/typeshare-cli)                                                                                                                                            |
+| typeshare            | [![crates.io version](https://img.shields.io/crates/v/typeshare.svg)](https://crates.io/crates/typeshare) [![crate documentation](https://docs.rs/typeshare/badge.svg)](https://docs.rs/typeshare) ![MSRV](https://img.shields.io/badge/rustc-1.75+-blue.svg) [![crates.io downloads](https://img.shields.io/crates/d/typeshare.svg)](https://crates.io/crates/typeshare)                                                                   |
```
[the remark this answers is not in the document]
Comment (darrell-roberts) [reply README.md]: Updated

**Review by CerulanLumina — APPROVED** [review]
Thank you for your hard work on this!

**Comment by CerulanLumina** [comment]
One request: would you mind rebasing to squash some commits for a cleaner history? 

There's also some merge conflicts that need resolving before I can merge.

Thanks again, and we'll get this merged this week!

**Pull request merged**